client/app/controllers/collection/add.js

```diff
@@ -83,43 +11,125 @@ export default Ember.Controller.extend({
     }),
 
     widgets: [],
+    actions: [],
+
+    // Fire enabled actions.
+    updateState: function(actions) {
+        actions.forEach((action) => {
+            // Check if the action can fire.
+            const may_fire = check_all.call(this, action.conditions);
+            if (!may_fire) return;
+            // Action may fire if execution has reached this point.
+            // Call the action and set its result and any
+            // changes to its state on `controller.parameters`.
+            console.log('stop here');
+            action.output_parameter['value'] = action.action.apply(this, action.arg_arr);
+            action.output_parameter['state'] = ['defined'];
+        });
+    },
+
+
+    // Take the description of an action and set its properties to be the vaious literal
+    // functions and parameters it depends on to operate.
+    hydrate_action: function(action) {
+        const parameters = this.get('parameters');
+        if (typeof parameters[action.parameter] !== 'object') {
+            parameters[action.parameter] = {};
+        }
+        // Create a new object as not to modify the object returned from the model
+        const hydrated_action = {
+            id: action.id,
+            type: action.type,
+            signature: this.get(action.type + '_signature'),
+            action: this.get(action.type),
+            conditions: action.conditions,
+            parameters: action.parameters,
+            args: action.args,
+            output_parameter: parameters[action.parameter]
+        };
+        var argarr = cons_arg_arr.call(this, hydrated_action);
+        hydrated_action['arg_arr'] = argarr
+        return hydrated_action;
 
-    create_widget(parameters) {
-        this.get('widgets').pushObject(parameters);
-        return parameters;
     },
 
-    run_update: function () {
-        updateState.call(this);
+    create_widget_signature: ['widget_component', 'description',
+                                        'section', 'output_parameter', 'action_id'],
```

Comment:
Signature allows fns to be curried before being passed into the component as a closure action, while remaining an obj on the model defn.